feat: add the `recordedtime` extension attribute

aucampia opened the pull request:
This extension defines an attribute that represents the time when an occurrence was recorded in a particular event, i.e. the time when the CloudEvent was created by a producer.

This enables bitemporal data to be represented with CloudEvents which enables several use cases described in the specification of the extension attribute.

Fixes #1200

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- Add the `recordedtime` extension attribute.

**Release Note**

Extension attributes are not tracked in the release notes as far as I know, so I'm eliding this. If it is needed let me know.

aucampia commented:
This is an early draft and I'm happy to make substantial changes to it.

duglin commented on cloudevents/extensions/recordedtime.md:
```diff
@@ -0,0 +1,106 @@
+
+- Type: `Timestamp`
+- Description: Timestamp of when the occurrence was recorded in this CloudEvent,
+  i.e. when the CloudEvent was created by a producer. If this attribute is used
```
I think the 2nd sentence is unneeded and could be misleading because it implies that if `recordedtime` isn't used then `time` might mean something different. Or are you just trying to say: "If this attribute is used then `time` SHOULD be present in the CloudEvent"  ?? I think that would be ok.

aucampia replied:
The CloudEvent spec says:

> If the time of the occurrence cannot be determined then this attribute MAY be set to some other time (such as the current time) by the CloudEvents producer

The second sentence here is an attempt to negate that. However, it is probably best to omit it, it really does not add much value and could add negative value.

aucampia replied:
I removed this now.

aucampia replied:
Removed it now.

duglin commented on cloudevents/extensions/recordedtime.md:
```diff
@@ -0,0 +1,106 @@
+## Usage
+
+When this extension is used, producers MUST set the value of the `recordedtime`
+attribute to the timestamp of when they create a new CloudEvent.
```
s/create a new/created the owning/

Something about the phrase "create a new" that feels like it could be interpreted as "some other CE rather the one with the recordedtime".

aucampia replied:
Changed.

duglin commented on cloudevents/extensions/recordedtime.md:
```diff
@@ -0,0 +1,106 @@
+
+## Use cases
+
+Examples of when an occurrence may need to be recorded differently are:
```
s/may/might/ since the "may" here isn't meant to be normative.

aucampia replied:
I changed this and a couple of other places where I think there were similar problems.

duglin commented on cloudevents/extensions/recordedtime.md:
```diff
@@ -0,0 +1,106 @@
+The recorded time also makes it easier to differentiate different records of the
+same occurrence in analytical data stores.
+
+## Terminology
```
I'd prefer if we didn't duplicate these terms here because then we need to worry about them getting out of sync.

aucampia replied:
I removed this and just link to the core spec, also happy to remove that, not entirely sure if it is needed, it is somewhat implied already.

duglin replied:
I don't think this section is needed at all since you already define "recordedtime" above.

aucampia replied:
Removing it, it does not really add much value.

duglin commented:
From the spec checker:
```
cloudevents/extensions/recordedtime.md: cloudevents/extensions/recordedtime.md does not contain '#event-data' segment
cloudevents/extensions/recordedtime.md: cloudevents/extensions/recordedtime.md does not contain '#context' segment
cloudevents/extensions/recordedtime.md: line 52: 'may' MUST be capitalized ('MAY')
cloudevents/extensions/recordedtime.md: line 57: 'may' MUST be capitalized ('MAY')
cloudevents/extensions/recordedtime.md: line 58: 'should' MUST be capitalized ('SHOULD')
cloudevents/extensions/recordedtime.md: Translation file cloudevents/languages/zh-CN/extensions/recordedtime.md does not exist
cloudevents/extensions/recordedtime.md: Translation file cloudevents/languages/he/extensions/recordedtime.md does not exist
```
For the translation files errors, just use https://github.com/cloudevents/spec/blob/main/cloudevents/languages/zh-CN/extensions/sequence.md and https://github.com/cloudevents/spec/blob/main/cloudevents/languages/he/extensions/sequence.md as guides to add placeholders.

duglin commented:
Overall it looks nice to me - thanks

aucampia commented:
@duglin thanks for the review. The CI passes now, the commit is signed and signed off, and I have addressed all the comments.

Not entirely sure what is up with GitHub, but to see diff from last review, check: https://github.com/cloudevents/spec/compare/07199e0..683f65f

duglin commented on cloudevents/extensions/recordedtime.md:
```diff
@@ -0,0 +1,83 @@
+- Description: Timestamp of when the occurrence was recorded in this CloudEvent,
+  i.e. when the CloudEvent was created by a producer.
+- Constraints:
+  - OPTIONAL
```
Mark this as REQUIRED. Extensions in general are optional, but once an extension is used this attribute is required.

aucampia replied:
I changed it to required.

duglin commented:
Just a few minor things, still LGTM

aucampia commented:
> Just a few minor things, still LGTM

Thanks, changes are made.

jskeet commented on cloudevents/extensions/recordedtime.md:
```diff
@@ -0,0 +1,68 @@
+  i.e. when the CloudEvent was created by a producer.
+- Constraints:
+  - REQUIRED
+  - If present, MUST adhere to the format specified in
```
Is there a constraint that this must be "at or later than" the occurrence time? Logically that would make sense, but if there are different clocks involved, I could imagine it not actually being strictly true in all cases.

aucampia replied:
I would say it should be later than the occurrence time, but as you say, it may be that there are different clocks involved, and the system creating CE may only be able to trust its own clock and may not want to assert this constraint, so I would not say it must be.

I will add this to the constraints:

```
- SHOULD be at or later than the _occurrence time_.
```

duglin commented:
@aucampia just one open comment for you to address and then I think we can vote on it

aucampia commented:
> @aucampia can you sign all of your commits? Or squash down to one and sign that one

I signed off on it now, and I added it to my repo config, so I won't forget again.

duglin commented:
thanks LGTM

duglin commented:
Approved on the 5/18 call
